File: Digital Business Enablement/bluegather/src/main/java/br/com/fiap/bluegather/service/AvaliacaoService.java
package br.com.fiap.bluegather.service;

import br.com.fiap.bluegather.dto.AvaliacaoDTO;
import br.com.fiap.bluegather.model.Avaliacao;
import br.com.fiap.bluegather.repository.AvaliacaoRepository;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.server.ResponseStatusException;

import java.util.Objects;

@Service
public class AvaliacaoService {

    @Autowired
    private AvaliacaoRepository avaliacaoRepository;

    @Autowired
    private EventoService eventoService;

    @Autowired
    private UsuarioService usuarioService;

    public Page<AvaliacaoDTO> listAll(Pageable pageRequest) {
        Page<Avaliacao> list = avaliacaoRepository.findAll(pageRequest);
        return list.map(this::convertToDto);
    }

    public AvaliacaoDTO findById(Long id) {
        Avaliacao entity = findEntityById(id);
        return convertToDto(entity);
    }

    @Transactional
    public AvaliacaoDTO create(AvaliacaoDTO newData) {
        Avaliacao entity = convertToEntity(newData);
        Avaliacao savedEntity = avaliacaoRepository.save(entity);
        return convertToDto(savedEntity);
    }

    @Transactional
    public AvaliacaoDTO update(Long id, AvaliacaoDTO updatedData) {
        findEntityById(id);
        updatedData.setId(id);
        Avaliacao updatedEntity = convertToEntity(updatedData);    
        Avaliacao savedEntity = avaliacaoRepository.save(updatedEntity);
        return convertToDto(savedEntity);
    }
    
    @Transactional
    public void delete(Long id) {
        Avaliacao entity = findEntityById(id);
        avaliacaoRepository.delete(entity);
    }

    public Avaliacao findEntityById(Long id) {
        return avaliacaoRepository.findById(id)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "(" + getClass().getSimpleName() + ") - Avaliacao não encontrado(a) por ID: " + id));
    }
    
    public AvaliacaoDTO convertToDto(Avaliacao entity) {
        AvaliacaoDTO dto = new AvaliacaoDTO();
        dto.setId(entity.getId());
        dto.setIdEvento(entity.getEvento() != null ? entity.getEvento().getId() : null);
        dto.setIdAvaliador(entity.getAvaliador() != null ? entity.getAvaliador().getId() : null);
        dto.setNota(entity.getNota());        
        return dto;
    }

    public Avaliacao convertToEntity(AvaliacaoDTO dto) {
        if (Objects.isNull(dto)) {
            throw new IllegalArgumentException("(" + getClass().getSimpleName() + ") - AvaliacaoDTO não pode ser nulo.");
        }
        Avaliacao entity;
        if (dto.getId() != null) {
            entity = findEntityById(dto.getId());
        } else {
            entity = new Avaliacao();
        }
        if (dto.getIdEvento() == null) {
            throw new IllegalArgumentException("(" + getClass().getSimpleName() + ") - idEvento não pode ser nulo.");
        }
        entity.setEvento(eventoService.findEntityById(dto.getIdEvento()));
        if (dto.getIdAvaliador() == null) {
            throw new IllegalArgumentException("(" + getClass().getSimpleName() + ") - idAvaliador não pode ser nulo.");
        }
        entity.setAvaliador(usuarioService.findEntityById(dto.getIdAvaliador()));
        entity.setNota(dto.getNota());       
        return entity;
    }
}

File: Enterprise Application Development/bluegather/Services/AvaliacaoService.cs
using Microsoft.EntityFrameworkCore;
using BlueGather.Dto;
using BlueGather.Models;
using BlueGather.Context;

namespace BlueGather.Services;
public class AvaliacaoService
{
    private readonly BlueGatherContext _context;

    public AvaliacaoService(BlueGatherContext context)
    {
        _context = context;
    }

    public async Task<List<AvaliacaoDto>> FindAll()
    {
        var list = await _context.Avaliacao
            .Include(x => x.Evento)
            .Include(x => x.Avaliador)
            .ToListAsync();

        return list.Select(entity => ConvertToDto(entity)).ToList();
    }

    public async Task<AvaliacaoDto> FindById(long id)
    {
        var entity = await FindEntityById(id);
        return ConvertToDto(entity);
    }

    public async Task<AvaliacaoDto> Create(AvaliacaoDto newData)
    {
        var entity = await ConvertToEntity(newData);
        _context.Avaliacao.Add(entity);
        await _context.SaveChangesAsync();
        return ConvertToDto(entity);
    }

    public async Task<AvaliacaoDto> Update(long id, AvaliacaoDto updatedData)
    {
        var entity = await FindEntityById(id);
        updatedData.Id = entity.Id;
        var updatedEntity = await ConvertToEntity(updatedData);
        _context.Entry(entity).CurrentValues.SetValues(updatedEntity);
        await _context.SaveChangesAsync();
        return ConvertToDto(entity);
    }

    public async Task Delete(long id)
    {
        var entity = await FindEntityById(id);
        _context.Avaliacao.Remove(entity);
        await _context.SaveChangesAsync();
    }

    public async Task<AvaliacaoModel> FindEntityById(long id)
    {
        var entity = await _context.Avaliacao
            .Include(x => x.Evento)
            .Include(x => x.Avaliador)
            .FirstOrDefaultAsync(x => x.Id == id);
        if (entity == null)
        {
            throw new KeyNotFoundException($"Objeto não encontrado com o ID: {id}");
        }
        return entity;
    }

    public async Task<List<AvaliacaoDto>> FindByEventoId(long id)
    {
        var list = await _context.Avaliacao
            .Where(x => x.Evento.Id == id)
            .Include(x => x.Evento)
            .Include(x => x.Avaliador)
            .ToListAsync();

        return list.Select(x => ConvertToDto(x)).ToList();
    }

    public async Task<List<AvaliacaoDto>> FindByAvaliadorId(long id)
    {
        var list = await _context.Avaliacao
            .Where(x => x.Avaliador.Id == id)
            .Include(x => x.Evento)
            .Include(x => x.Avaliador)
            .ToListAsync();

        return list.Select(x => ConvertToDto(x)).ToList();
    }

    private AvaliacaoDto ConvertToDto(AvaliacaoModel entity)
    {
        return new AvaliacaoDto
        {
            Id = entity.Id,
            IdEvento = entity.Evento.Id,
            IdAvaliador = entity.Avaliador.Id,
            Nota = entity.Nota
        };
    }

    private async Task<AvaliacaoModel> ConvertToEntity(AvaliacaoDto dto)
    {
        var evento = await _context.Evento.FindAsync(dto.IdEvento);
        if (evento == null)
        {
            throw new KeyNotFoundException($"Objeto não encontrado com o ID: {dto.IdEvento}");
        }

        var avaliador = await _context.Usuario.FindAsync(dto.IdAvaliador);
        if (avaliador == null)
        {
            throw new KeyNotFoundException($"Objeto não encontrado com o ID: {dto.IdAvaliador}");
        }

        return new AvaliacaoModel
        {
            Id = dto.Id ?? 0,
            Evento = evento,
            Avaliador = avaliador,
            Nota = dto.Nota
        };
    }
}